### PR TITLE
Add python equivalent for  SFMExample_bal.cpp

### PR DESCRIPTION
**Pull request by johnwlambert**
Adds Bundle Adjustment functionality to the Python wrapper.

Should provide the equivalent of the C++ example (https://github.com/borglab/gtsam/blob/develop/examples/SFMExample_bal.cpp)

Present ambiguity @dellaert  -- is
```c++
typedef gtsam::PinholeCamera<gtsam::Cal3Bundler> PinholeCameraCal3Bundler;
```
or
```c++
typedef gtsam::PinholeCamera<gtsam::Cal3Bundler> SfmCamera;
```
preferred? We use both throughout codebase currently.

Introduces:

- PriorFactor for SfmCamera
- GeneralSFMFactor<SfmCamera,Point3>
See Line 34 and Line 68 of https://github.com/borglab/gtsam/blob/develop/examples/SFMExample_bal.cpp

A later PR should include
- SfmData -> set up of collection is considered part of “initialization”. Can add .push_back() method to this struct, Unit tests of newly wrapped functionality

### CODE REVIEW

File: gtsam/gtsam.i [hunk 1]
```diff
@@ -1106,7 +1106,7 @@ typedef gtsam::PinholeCamera<gtsam::Cal3_S2> PinholeCameraCal3_S2;
 //TODO (Issue 237) due to lack of jacobians of Cal3DS2_Base::calibrate, PinholeCamera does not apply to Cal3DS2/Unified
 //typedef gtsam::PinholeCamera<gtsam::Cal3DS2> PinholeCameraCal3DS2;
 //typedef gtsam::PinholeCamera<gtsam::Cal3Unified> PinholeCameraCal3Unified;
-typedef gtsam::PinholeCamera<gtsam::Cal3Bundler> PinholeCameraCal3Bundler;
+typedef gtsam::PinholeCamera<gtsam::Cal3Bundler> SfmCamera;
```
Comment (dellaert) [inline gtsam/gtsam.i:1109]: Don’t globally replace, please. PinholeCameraCal3Bundler makes more sense. Only use SfmCamera in conjunction with the dataset.h data structures, which are hard coded for Cal3Bundler for now.

File: python/gtsam/examples/SFMExample_bal.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import argparse
+import logging
+import sys
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+import gtsam
+from gtsam import (
+    GeneralSFMFactorSfmCamera,
```
Comment (dellaert) [inline python/gtsam/examples/SFMExample_bal.py:22]: Please format entire file with google style

File: python/gtsam/examples/SFMExample_bal.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    PriorFactorSfmCamera,
+    readBal,
+    symbol_shorthand
+)
+
+C = symbol_shorthand.C
+P = symbol_shorthand.P
+
+
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
```
Comment (dellaert) [inline python/gtsam/examples/SFMExample_bal.py:33]: I like it!

File: python/gtsam/examples/SFMExample_bal.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+# We will be using a projection factor that ties a SFM_Camera to a 3D point.
+# An SfmCamera is defined in dataset.h as a pinhole camera with unknown Cal3Bundler calibration
+# and has a total of 9 free parameters
+
+def run(args):
+    """ Run LM optimization with BAL input data and report resulting error """
+    # Find default file, but if an argument is given, try loading a file
+    if args.input_file:
```
Comment (dellaert) [inline python/gtsam/examples/SFMExample_bal.py:42]: That can be done using argparse defaults below

**Review by dellaert — APPROVED** [review]
Nice, but please don’t do the SfMCamera name change

**Comment by johnwlambert** [comment]
@dellaert Hi Frank, should be ready for a re-review now. Thanks!

File: python/gtsam/examples/SFMExample_bal.py (hunk 2 above)
Comment (johnwlambert) [reply python/gtsam/examples/SFMExample_bal.py]: I'm following `black` here for the multi-line imports. Is there a particular PEP number you could point me to for Google style?

Comment (johnwlambert) [reply python/gtsam/examples/SFMExample_bal.py]: https://github.com/psf/black

**Comment by johnwlambert** [comment]
Hi @dellaert @akshay-krishnan, can we go ahead and merge this today?

@xwu4lab @swarrier246, it might be helpful for you all to see the factors we use here as well.

**Review by swarrier246 — COMMENTED** [review]
would it be helpful to mention what the input file contents and format will be?

File: python/gtsam/examples/SFMExample_bal.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+
+def run(args):
+    """ Run LM optimization with BAL input data and report resulting error """
+    input_file = gtsam.findExampleDataFile(args.input_file)
+
+    # Load the SfM data from file
+    mydata = readBal(input_file)
```
Comment (akshay-krishnan) [inline python/gtsam/examples/SFMExample_bal.py:40]: I think it would be nice to rename mydata to something more descriptive of the type of data.

**Review by akshay-krishnan — APPROVED** [review]

File: python/gtsam/examples/SFMExample_bal.py (hunk 5 above)
Comment (johnwlambert) [reply python/gtsam/examples/SFMExample_bal.py:40]: how about `scene_data` @akshay-krishnan ?

**Comment by johnwlambert** [comment]
@swarrier246 thanks for the review, just added an extended comment about what the `BAL` contents look lke

**Review by dellaert — APPROVED** [review]

**Pull request merged**